### PR TITLE
Om th

### PR DESCRIPTION
**Pull request by derekpappas**
**Related issue**: 

**Type of change**: feature request

**Impact**:  API addition (no impact on existing code)

**Development Phase**:  implementation

**Release Notes**

Adding support for logical tree nodes in diplomatic SRAM's
* This PR adds ResourceBindings to the test rams
* This PR does not remove the old omMem return value for now.
* Supports multiple roots - design and test harness roots are added under root now
* The Logical TreeNodes are passed down the DUT hierarchy through function parameters
* The DescribedSRAM now has a class and the object has two functions: apply and build
  * build is used by the test harness memories -Returns an object which has the information needed to construct an OMMemory - includes a hash function for matching the DescribedSRAM annotation that contains the same hash value. The DescribedSRAM anno will have the Veirlog path. The anno and the OM mem will be joined in a post process step. Another PR can add the hash to the design memories as well. 
  * apply is used by all legacy memories

### CODE REVIEW

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala [hunk 1]
```diff
@@ -23,7 +29,7 @@ object LogicalModuleTree {
 
   def root: LogicalTreeNode = {
     val roots = tree.collect { case (k, _) if !tree.exists(_._2.contains(k)) => k }
-    assert(roots.size == 1, "Logical Tree contains more than one root.")
+    assert(roots.size <= 2, "Logical Tree contains more than two roots.")
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: "two or more"?

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala [hunk 2]
```diff
@@ -11,7 +12,12 @@ trait LogicalTreeNode {
   def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent] = Nil): Seq[OMComponent]
 }
 
+class RootLogicalTreeNode extends LogicalTreeNode {
+def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent] = Nil): Seq[OMComponent] = children
+}
+
 object LogicalModuleTree {
+  val rootLTN = new RootLogicalTreeNode()
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: I'd just call this `root`

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 package freechips.rocketchip.diplomaticobjectmodel.logicaltree
 
+import freechips.rocketchip.config.Field
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: Unneeded import?

File: src/main/scala/amba/axi4/SRAM.scala [hunk 4]
```diff
@@ -15,7 +17,7 @@ class AXI4RAM(
     errors: Seq[AddressSet] = Nil,
     wcorrupt: Boolean = false)
   (implicit p: Parameters) extends DiplomaticSRAM(address, beatBytes, devName)
-{
+  {
```
Comment (hcook) [inline src/main/scala/amba/axi4/SRAM.scala]: Why did you change all the whitespace in this one file?

File: src/main/scala/amba/axi4/SRAM.scala [hunk 5]
```diff
@@ -5,6 +5,8 @@ package freechips.rocketchip.amba.axi4
 import Chisel._
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree._
+import freechips.rocketchip.diplomaticobjectmodel.model.{OMAPBRAM, OMAXI4RAM}
```
Comment (hcook) [inline src/main/scala/amba/axi4/SRAM.scala]: Unneeded import

File: src/main/scala/amba/apb/SRAM.scala [hunk 6]
```diff
@@ -5,6 +5,8 @@ package freechips.rocketchip.amba.apb
 import Chisel._
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.diplomaticobjectmodel.logicaltree._
+import freechips.rocketchip.diplomaticobjectmodel.model.{OMAPBRAM, OMTLRAM}
```
Comment (hcook) [inline src/main/scala/amba/apb/SRAM.scala]: Unneeded import

File: src/main/scala/amba/ahb/SRAM.scala [hunk 7]
```diff
@@ -5,6 +5,7 @@ package freechips.rocketchip.amba.ahb
 import Chisel._
 import freechips.rocketchip.config.Parameters
 import freechips.rocketchip.diplomacy._
+import freechips.rocketchip.diplomaticobjectmodel.model.{OMAHBRAM, OMTLRAM}
```
Comment (hcook) [inline src/main/scala/amba/ahb/SRAM.scala]: Unneeded import

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 8]
```diff
@@ -97,3 +99,71 @@ class SubSystemLogicalTreeNode(var getOMInterruptDevice: (ResourceBindingsMap) =
...
+
+case object DeviceKey extends Field[Option[Device]](
+  None
+)
+
+case object ParentLogicalTreeNodeKey extends Field[Option[LogicalTreeNode]](
+  None
+)
+
+case object ResourceBindingsMapLogicalTreeNodeKey extends Field[Option[ResourceBindingsMap]](
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: What does this have to do with "LogicalTreeNodes"? Also, why do you need it? Aren't the ResourceBindings supposed to be resolved when you walk the tree at the end?

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 9]
```diff
@@ -97,3 +99,71 @@ class SubSystemLogicalTreeNode(var getOMInterruptDevice: (ResourceBindingsMap) =
...
+  None
+)
+
+object ParentTreeNode {
+  /**
+    * Add the parent tree node to the Parameters so that the children can access the tree node via the params
+    * @param p
+    * @return
+    */
+  def set(pltnOpt: Option[LogicalTreeNode], ltn: LogicalTreeNode)(implicit p: Parameters): Parameters = {
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: Is this used?

File: src/main/scala/subsystem/BaseSubsystem.scala [hunk 10]
```diff
@@ -77,6 +77,7 @@ abstract class BaseSubsystem(implicit p: Parameters) extends BareSubsystem with
   def getOMInterruptDevice(resourceBindingsMap: ResourceBindingsMap): Seq[OMInterrupt]
 
   val logicalTreeNode = new SubSystemLogicalTreeNode(getOMInterruptDevice)
+  LogicalModuleTree.add(LogicalModuleTree.rootLTN, logicalTreeNode)
```
Comment (hcook) [inline src/main/scala/subsystem/BaseSubsystem.scala]: Should this really always be added to the root? What if there are other layers in between?

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 11]
```diff
@@ -97,3 +99,71 @@ class SubSystemLogicalTreeNode(var getOMInterruptDevice: (ResourceBindingsMap) =
...
+  data: T,
+  resourceBindings: ResourceBindings,
+  architecture: RAMArchitecture,
+  description: String,
+  depth: Int,
+  ecc: Option[OMECC] = None,
+  hasAtomics: Option[Boolean] = None
+)
+
+class SRAMLogicalTreeNode[T <: Data](description: String, architecture: RAMArchitecture, depth: Int, lanes: Int, bits: Int, device: () => Device,
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: Why does this class need both `rbm: () => ResourceBindingsMap` AND `def getOMComponents(resourceBindingsMap: ResourceBindingsMap,...)` Are these the same RBM, or different RBMs? Something seems wrong about the fact that there are two and you only use one...

File: src/main/scala/diplomaticobjectmodel/ConstructOM.scala [hunk 12]
```diff
@@ -7,9 +7,14 @@ import freechips.rocketchip.diplomaticobjectmodel.logicaltree._
 import freechips.rocketchip.diplomaticobjectmodel.model.OMComponent
 import freechips.rocketchip.util.ElaborationArtefacts
 
-case object ConstructOM {
-  def constructOM(resourceBindingsMap: => ResourceBindingsMap): Unit = {
-    val om: Seq[OMComponent] = LogicalModuleTree.bind(resourceBindingsMap)
+object ConstructOM {
+  private var resourceBindingsMap: () => ResourceBindingsMap = _
+
+  def apply(rbm: () => ResourceBindingsMap): Unit = resourceBindingsMap = rbm
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/ConstructOM.scala]: I'd call this method `bind` or `setBindingsMap` or something more explicit than just the apply method. Usually when object have an apply method they return an instance of the companion class.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 13]
```diff
@@ -97,3 +99,71 @@ class SubSystemLogicalTreeNode(var getOMInterruptDevice: (ResourceBindingsMap) =
...
+    children
+}
+
+class TestSoCLogicalTreeNode extends LogicalTreeNode {
+  override def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent] = Nil): Seq[OMComponent] = {
+    Seq(OMTestHarness( components = children))
+  }
+}
+
+case object DeviceKey extends Field[Option[Device]](
```
Comment (hcook) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: What is this for?

File: src/main/scala/diplomacy/Resources.scala [hunk 14]
```diff
@@ -183,11 +183,14 @@ class SimpleDevice(devname: String, devcompat: Seq[String]) extends Device
   with DeviceClocks
   with DeviceRegName
 {
+  var resourceBindings: Option[ResourceBindings] = None
```
Comment (hcook) [inline src/main/scala/diplomacy/Resources.scala]: @terpstra Are devices supposed to have their own ResourceBindings?

File: src/main/scala/util/DescribedSRAM.scala [hunk 15]
```diff
@@ -29,6 +24,8 @@ object DescribedSRAM {
       case d => d.getWidth
     }
 
+    val f = () => mem.toNamed.toString.hashCode
```
Comment (hcook) [inline src/main/scala/util/DescribedSRAM.scala]: What is this for? Where is it used?

File: src/main/scala/diplomaticobjectmodel/ConstructOM.scala (hunk 12 above)
Comment (derekpappas) [reply src/main/scala/diplomaticobjectmodel/ConstructOM.scala]: Thanks.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala (hunk 1 above)
Comment (derekpappas) [reply src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: The structure of the tree is as follows: @terpstra @richardxia 

                      root
                       /   \
                     /       \
TestHarness        CoreComplex

Comment (richardxia) [reply src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: I don't really have much context for what is happening right now, but if we are already going from 1 to 2 roots, is there any downside to just allowing an unbounded number of roots? We essentially already have a forest.

File: src/main/scala/diplomaticobjectmodel/ConstructOM.scala (hunk 12 above)
Comment (derekpappas) [reply src/main/scala/diplomaticobjectmodel/ConstructOM.scala]: Changed to:

  def setBindingsMap(rbm: () => ResourceBindingsMap): Unit = resourceBindingsMap = rbm

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala [hunk 16]
```diff
@@ -97,3 +101,53 @@ class SubSystemLogicalTreeNode(var getOMInterruptDevice: (ResourceBindingsMap) =
...
+        memories = Nil,
+        architecture = ramArchitecture
+      )
+    )
+  }
+
+  def getOMComponents(resourceBindingsMap: ResourceBindingsMap, components: Seq[OMComponent]): Seq[OMComponent] = {
+    val c = DiplomaticObjectModelAddressing.getOMComponentHelper(device(), resourceBindingsMap, getOMTestMemory)
+    val cc = components.map(_.asInstanceOf[OMMemory])
+    val children = c.toList.head.asInstanceOf[OMTestMemory].copy( memories = cc).asInstanceOf[OMComponent]
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTrees.scala]: This feels awkward, having to take the `c` object, which we just happen to know is a Seq containing exactly one `OMTestMemory` and having to grab the first element and cast it to an `OMTestMemory`. I know that the type system only knows that `c` is a Seq of `OMComponent`s, but this seems like we are throwing away type safety.

I think there's a way to make this work without explicit type casts by adding a type parameter to `getOMComponentHelper`, but I don't know if that refactoring is a good idea here.

File: src/main/scala/tilelink/SRAM.scala [hunk 17]
```diff
@@ -201,8 +207,8 @@ class TLRAM(
       if (code.canDetect) code.encode(d, p) else code.encode(d)
     })
 
-    d_raw_data := mem.read(addr, ren)
-    when (wen) { mem.write(addr, coded, sel) }
+    d_raw_data := mem.mem.read(addr, ren)
```
Comment (richardxia) [inline src/main/scala/tilelink/SRAM.scala]: This looks a little bit awkward and confusing now, with the double `mem`. Could we give the outer `mem` a different name up on line 61 here?

This kind of change will need to be replicated to all the other SRAM instances for the other protocols.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala [hunk 18]
```diff
@@ -11,27 +12,47 @@ trait LogicalTreeNode {
   def getOMComponents(resourceBindingsMap: ResourceBindingsMap, children: Seq[OMComponent] = Nil): Seq[OMComponent]
 }
 
+class RootLogicalTreeNode extends LogicalTreeNode {
```
Comment (derekpappas) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: Make one concrete generic LTN and replace this LTN and the ImpLTN with the single concrete LTN.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala [hunk 19]
```diff
@@ -11,27 +12,47 @@ trait LogicalTreeNode {
...
-  def add(parent: LogicalTreeNode, child: => LogicalTreeNode): Unit = {
+
+  def addResourceScope(ltn: LogicalTreeNode, rbm: () => ResourceBindingsMap): Unit = {
+    require(! resourceScopes.contains(ltn))
+    resourceScopes.put(ltn, rbm)
+  }
+
+  def add(parent: LogicalTreeNode, child: => LogicalTreeNode, resourceBindingsMap: Option[() => ResourceBindingsMap] = None): Unit = {
     val treeOpt = tree.get(parent)
     val treeNode = treeOpt.map{
```
Comment (derekpappas) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: Add a check that the parent does not contain the child already.

Add a check that the child has not been added to the tree more than once.

This can be done in the map add function.

This means when adding a parent child edge in the map that the child should not appear in any other parent child list. This can be checked with a second map of all children.

File: src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala [hunk 20]
```diff
@@ -11,27 +12,47 @@ trait LogicalTreeNode {
...
     roots.head
   }
 
   def bind(resourceBindingsMap: ResourceBindingsMap): Seq[OMComponent] = {
     def getOMComponentTree(node: LogicalTreeNode): Seq[OMComponent] = {
-      node.getOMComponents(resourceBindingsMap, tree.get(node).getOrElse(Nil).flatMap(getOMComponentTree))
+      val rbm = resourceScopes.get(node) match {
+        case Some(r) =>
+          r()
+        case _ => resourceBindingsMap
```
Comment (derekpappas) [inline src/main/scala/diplomaticobjectmodel/logicaltree/LogicalTreeNode.scala]: Change to case None

File: src/main/scala/amba/ahb/Parameters.scala [hunk 21]
```diff
@@ -28,6 +28,15 @@ case class AHBSlaveParameters(
...
   // The device had better not support a transfer larger than it's alignment
   require (minAlignment >= maxTransfer)
+
+  def toResource: ResourceAddress = {
+    ResourceAddress(address, ResourcePermissions(
+      r = supportsRead,
+      w = supportsWrite,
+      x = executable,
+      c = false,
+      a = false))
```
Comment (derekpappas) [inline src/main/scala/amba/ahb/Parameters.scala]: @terpstra @richardxia wants to know if the default = false is always true
